src/core/internal/gc/impl/conservative/gc.d

```diff
@@ -2648,29 +2668,50 @@ struct Gcx
         // (unless they allocate or use the GC themselves, in which case
         // the global GC lock will stop them).
         // fork now and sweep later
+        int child_mark() scope
+        {
+            if (doParallel)
+                markParallel(nostack);
+            else if (ConservativeGC.isPrecise)
+                markAll!(markPrecise!true)(nostack);
+            else
+                markAll!(markConservative!true)(nostack);
+            return 0;
+        }
+
         import core.stdc.stdlib : _Exit;
         debug (PRINTF_TO_FILE)
         {
             import core.stdc.stdio : fflush;
             fflush(null); // avoid duplicated FILE* output
         }
-        fork_needs_lock = false;
-        auto pid = fork();
-        fork_needs_lock = true;
+        static if (has_clone)
+        {
+            const flags = CLONE_CHILD_CLEARTID | SIGCHLD; // child thread id not needed
+            scope int delegate() scope dg = &child_mark;
+            extern(C) static int wrap_delegate(void* arg)
+            {
+                auto dg = cast(int delegate() scope*)arg;
+                return (*dg)();
+            }
+            char[256] stackbuf; // enough space to place some info for the child without stomping the parent stack
+            auto stack = stackbuf.ptr + (isStackGrowingDown ? stackbuf.length : 0);
+            auto pid = clone(&wrap_delegate, stack, flags, &dg);
+        }
+        else
+        {
+            fork_needs_lock = false;
+            auto pid = fork();
+            fork_needs_lock = true;
```

Comment:
Does it makes sense to keep supporting a fork-based approach if it's known to be buggy and lead to deadlock? Maybe look for the existence of `_Fork` instead so any OS adopting this new function can be supported too?

Reply:
I'm not so sure either, but it is the version that is also tested for FreeBSD and that passes, too. The deadlock might not manifest on all platforms and probably also depends on the actual C library used.

`_Fork` is just being released in libc, so it might take some time to be available on all platforms.